ttrss_api: Fix relogin

noctux opened the pull request:
The error syntax returned by TTRSS if their session is invalidated looks like this:

```json
{"seq":null,"status":1,"content":{"error":"NOT_LOGGED_IN"}}
```

A bug introduced while switching to json.hpp caused newsboat to look for the "error" component in the toplevel object, and not within the "content" subobject, which is fixed in this patch.

Once more, mea culpa... I think I definitely have to look into getting some kind of test coverage into the ttrss-api... I'm sorry :/

noctux commented:
Hmm, not exactly the point of this PR, but how exactly is coverage measured?
I've moved a block of code within the method and changed the accessed object. But on a "number of operations of type X" scale, the code remained stable, so I'm really intrigued in how this results into an 13% decrease in coverage...

tsipinakis commented:
> I'm really intrigued in how this results into an 13% decrease in coverage...

Correct me if I'm wrong but I think it's because the branch you're trying to merge is quite a bit behind master and coveralls apparently compares the coverage results against master and not the merge-base (read: commit you branched out from).

So if your PR doesn't alter coverage at all and meanwhile a commit is pushed to master than increases it by 13% when it tries to compare the two it'll count it as a 13% drop.

noctux commented:
> Correct me if I'm wrong but I think it's because the branch you're trying to merge is quite a bit behind master and coveralls apparently compares the coverage results against master and not the merge-base (read: commit you branched out from).

Ah, yes, indeed. As it seems that final rebase was only local and I forgot to force-push to my github branch... Thanks for having a look into it!

Minoru commented:
Thanks!

Re tests: I gave it some thought a while back, and couldn't decide what approach to use in order to get any benefit. Unit testing would've caught this particular bug, but won't help with keeping the API up to date. Integration testing against real-world servers would help with updating the API, but I'm not sure if it's even doable—I wager remote APIs won't serve us the same content each time, so it's not clear what exactly such a test suite will check.

But then, I don't have any experience with tests other than unit testing. Maybe I'm missing something?

noctux commented:
Thanks for merging!

Well, I was just reconsidering the switch to json.hpp: I've introduced (at least) two subtle bugs (the attachment typo: #177 and this one), which went unnoticed to a large degree, because I personally didn't use these functionalities too intensely. Considering that we are pondering to move other apis as well, a smoother porting experience might be desireable...

But there is another problem: For most remote apis, we do not have any idea what the responses across different api-versions look like, so if things start to break, we always have to consider two possibilities: The api has changed, or we broke something. Having tests for all supported functionality always gives us the ability to diff the actual api-responses to the ones we saw when we implemented the api.

So I've given the idea some thought on how one could implement clean tests, but did not arrive at any good conclusion. It seems like we would have two options, one being simpler but a lot more hackish, and the other being cleaner, but rather difficult. Both center on the idea of having "known good" communication traces, which we check against:
1) Most APIs use `utils::retrieve_url`. We could use objcopy to weaken that symbol in the .o file when linking to the tests, and tests concering the remote api could implement their own version, which hands out predefined results. Definitely hackish, but somewhat easy to do.
2) Implementing a server stub that stores different communication traces with the remote and hands those out... that way, we test against a know good server. But spawning the server, asserting that it's there, communicating which trace is relevant, etc. sounds like a major hassle.

So overall, I would really like to have at least such a limited degree of tests to have a more confident experience when refactoring and a "known good" diff for troubleshooting when the remote api changes...

But maybe we should break this discussion into its own issue...

Minoru commented:
> But maybe we should break this discussion into its own issue...

Definitely; I created #190, let's move there.
